### PR TITLE
Rework connection settings menu system

### PR DESCRIPTION
**Pull request by abrooksv**
* Decouple the connection settings menu's actions from a Project
* Make "Recent X" options optional
* Introduce a label combo action for use in UIs
* Allow connection settings to be project level or local (for UIs)

Draft: Few more test cases need to be written for the local and project state classes, will finish them next week

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenus.kt [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    fun displayValue(): String = when (menuMode) {
+        CREDENTIALS -> credentialsDisplay()
+        REGIONS -> regionDisplay()
+        BOTH -> "${credentialsDisplay()}@${regionDisplay()}"
+    }
+
+    fun tooltip(): String? = when (menuMode) {
+        CREDENTIALS -> credentialsTooltip()
+        REGIONS -> regionTooltip()
+        BOTH -> null
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenus.kt:63]: am surprised this lacks a tooltip

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        selected: Boolean,
+        onSelect: (CredentialIdentifier) -> Unit
+    ) : DumbAwareToggleAction<CredentialIdentifier>(value.displayName, value, selected, onSelect)
+
+    companion object {
+        private const val ID = "ConnectionSettingsMenu"
+
+        fun connectionSettingsMenuBuilder(): ConnectionSettingsMenuBuilder = ConnectionSettingsMenuBuilder()
+
+        fun projectConnectionSettingsMenu(project: Project, dataContext: DataContext, showRegions: Boolean = true, showCredentials: Boolean = true): ListPopup {
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt:168]: feels like we could use ProjectLevelSettingSelector here?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+class ConnectionSettingsMenuBuilder private constructor() {
+    private data class RegionSelectionSettings(val currentSelection: AwsRegion?, val onChange: (AwsRegion) -> Unit)
+    private data class CredentialsSelectionSettings(val currentSelection: CredentialIdentifier?, val onChange: (CredentialIdentifier) -> Unit)
+
+    private var regionSelectionSettings: RegionSelectionSettings? = null
+    private var credentialsSelectionSettings: CredentialsSelectionSettings? = null
+    private var accountSettingsManager: AwsConnectionManager? = null
+
+    fun withRegions(currentSelection: AwsRegion?, onChange: (AwsRegion) -> Unit): ConnectionSettingsMenuBuilder {
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt:32]: nit: `apply {}`?

Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt:32]: nit: `withRegions()` and `withRegionSelected(currentSelection)`? not sure how you'd clarify that it'd be either/or

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt:32]: Oh, thats nicer. Thx for reminder

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenus.kt (hunk 1 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenus.kt:61]: Ya, current one does so it was kept that way. Not sure how I feel about it though...

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt (hunk 2 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt:168]: Good point, it predates me building ProjectLevelSettingSelector. Moved to it.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt]: Not sure there is a great way to "solve" that without interface / type safe builders which is a lot of boilerplate for this.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ChangeConnectionSettingsMenu.kt [hunk 4]
```diff
@@ -1,232 +0,0 @@
...
-
-            getAccountSetting(e).changeRegion(region)
-
-            if (oldRegion?.partitionId != region.partitionId) {
-                AwsTelemetry.setPartition(
-                    partitionId = region.partitionId
-                )
-            }
-
-            AwsTelemetry.setRegion(e.project)
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ChangeConnectionSettingsMenu.kt:172]: need these back

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ChangeConnectionSettingsMenu.kt:172]: Yikes...good catch

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ChangeConnectionSettingsMenu.kt:172]: Moved logic into AwsConnectionManager which matches where we do the validate telemetry

**Review by rli — APPROVED** [review]

**Pull request merged**